app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/miscale2/MiScale2DeviceSupport.java

```diff
@@ -0,0 +1,290 @@
+        builder.add(new SetDeviceStateAction(getDevice(), GBDevice.State.INITIALIZED, getContext()));
+
+        // Weight and body composition
+        builder.setGattCallback(this);
+        builder.notify(getCharacteristic(GattCharacteristic.UUID_CHARACTERISTIC_BODY_COMPOSITION_MEASUREMENT), true);
+
+        return builder;
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
+        super.onCharacteristicChanged(gatt, characteristic);
+
+        UUID characteristicUUID = characteristic.getUuid();
+        if (characteristicUUID.equals(GattCharacteristic.UUID_CHARACTERISTIC_BODY_COMPOSITION_MEASUREMENT)) {
+            final byte[] data = characteristic.getValue();
+
+            boolean stabilized = testBit(data[1], 5) && !testBit(data[1], 7);
+            boolean isLbs = testBit(data[1], 0);
+            boolean isJin = testBit(data[1], 4);
+            boolean isKg = !(isLbs && isJin);
+            String unit = "";
+            if (isKg) {
+                unit = UNIT_KG;
+            } else if (isLbs) {
+                unit = UNIT_LBS;
+            } else if (isJin) {
+                unit = UNIT_JIN;
+            }
+
+            if (stabilized) {
+                int year = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, 2);
+                int month = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 4);
+                int day = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 5);
+                int hour = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 6);
+                int minute = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 7);
+                int second = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 8);
+                Calendar c = GregorianCalendar.getInstance();
+                c.set(year, month - 1, day, hour, minute, second);
+                Date date = c.getTime();
+                float weight = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, 11) / (isKg ? 200.0f : 100.0f);
+                handleWeightInfo(date, weight, unit);
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean testBit(byte value, int offset) {
+        return ((value >> offset) & 1) == 1;
+    }
+
+    private void handleDeviceInfo(nodomain.freeyourgadget.gadgetbridge.service.btle.profiles.deviceinfo.DeviceInfo info) {
+        LOG.warn("Device info: " + info);
+        versionCmd.hwVersion = info.getHardwareRevision();
+        versionCmd.fwVersion = info.getSoftwareRevision();
+        handleGBDeviceEvent(versionCmd);
+    }
+
+    private void handleWeightInfo(Date date, float weight, String unit) {
+        // TODO
+        LOG.warn("Weight info: " + weight + unit);
+        GB.toast(weight + unit, Toast.LENGTH_SHORT, GB.INFO);
```

Comment:
Here we need to think how to handle this UI-wise and data-wise, I guess.

We should probably store the results in the db. And fire an event so that the UI can display the current value and possibly a chart with old values.

Reply:
The amount of changes were a bit overwhelming to me :)
Yes, store a body information in DB is probably the next step